src/main/java/com/codecool/shop/controller/ProductController.java

```diff
@@ -2,39 +2,211 @@
 public class ProductController extends HttpServlet {
 
+    private StringBuffer categoryNames = new StringBuffer();
+    private StringBuffer supplierNames = new StringBuffer();
+    private boolean requestForCategory = false;
+    private boolean requestForSupplier = false;
+
     @Override
     protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+
+        categoryNames.delete(0, categoryNames.length());
+        supplierNames.delete(0, supplierNames.length());
+        requestForSupplier = false;
+        requestForCategory = false;
+
         ProductDao productDataStore = ProductDaoMem.getInstance();
         ProductCategoryDao productCategoryDataStore = ProductCategoryDaoMem.getInstance();
+        SupplierDao supplierDataStore = SupplierDaoMem.getInstance();
+
+        List<Product> products = filterProduct(req, supplierDataStore, productCategoryDataStore, productDataStore);
 
         TemplateEngine engine = TemplateEngineUtil.getTemplateEngine(req.getServletContext());
+
         WebContext context = new WebContext(req, resp, req.getServletContext());
-        context.setVariable("category", productCategoryDataStore.find(1));
-        context.setVariable("products", productDataStore.getBy(productCategoryDataStore.find(1)));
-        // // Alternative setting of the template context
-        // Map<String, Object> params = new HashMap<>();
-        // params.put("category", productCategoryDataStore.find(1));
-        // params.put("products", productDataStore.getBy(productCategoryDataStore.find(1)));
-        // context.setVariables(params);
+        addVariablesToContext(context, productCategoryDataStore, products, supplierDataStore);
+
         engine.process("product/index.html", context, resp.getWriter());
     }
 
+    private void addVariablesToContext(WebContext context, ProductCategoryDao productCategoryDataStore,
+                                       List<Product> products, SupplierDao supplierDataStore){
+        context.setVariable("allCategory", productCategoryDataStore.getAll());
+        context.setVariable("allSupplier", supplierDataStore.getAll());
+
+        context.setVariable("categories", categoryNames.toString().trim().replaceAll(",$", ""));
+        context.setVariable("suppliers", supplierNames.toString().trim().replaceAll(",$", ""));
+        context.setVariable("products", products);
+
+        context.setVariable("cartListLength", Cart.getCartListSize());
+    }
+
+    /**
+     * This method will get all products in the given categories and sets the boolean for request true, if there were any.
+     *
+     * @param request                  http request
+     * @param productCategoryDataStore stores the categories
+     * @param productDataStore         stores the products
+     * @return the filtered product List
+     */
+    private List<Product> filterProductsByGivenCategories(HttpServletRequest request, ProductCategoryDao productCategoryDataStore,
```

Comment:
Actually this part is quite heavy so I would extract the logic into a service e.g. `CategoryFilterService` which encapsulates this complex logic.